Ipld gitbookv2 (#111)
protocol/launchpad: docs/ipld/content-addressing-and-cids.md

* How many bytes of output we have
* What kind of data is being addressed and how we might interpret it when we find it

> ### CIDs are immutable Links

Reviewer: ```suggestion
### CIDs are Immutable Links
```
?

### Multiformats

> [Multiformats](https://multiformats.io/) were created to make it easy to switch data formats on IPFS (and [any other system that implements IPLD](https://multiformats.io/#projects-using-multiformats)), future proofing them by making them self-describing, which prevents breaking changes and make it possible for  to anything that uses that data format to interoperate with (or upgrade to) other systems that use differently defined data.

Reviewer: My main concern with this is the multiformats is much broader, it's a system for self-describing "values" (where "values" is somewhat broad). We use it for IPLD for codecs and hash function definition in CIDs, but it's also used in IPFS for protocol/address encoding (not my field), and it has a number of other novel use-cases outside of IPLD and even IPFS. So, attempting to broad hand-wave and narrow down to IPLD:

```suggestion
[Multiformats](https://multiformats.io/) were created to make formats, protocols, hash digests and other small values self-describing. They are useful in IPFS (and [any other system that implements IPLD](https://multiformats.io/#projects-using-multiformats)) for communicating encoding formats (codecs) and hash functions. Their self-describing nature is a form of future-proofing which prevents breaking changes and allows for interoperability between a wide variety hash functions and encoding formats, including those from other content-addressed systems.
```

#### Multiformat Protocols

> * **Multihash**: Self-describing hashes
> * **Multiaddr**: Self-describing network addresses
> * **Multibase**: Self-describing base encodings
> * **Multicodec**: Self-describing serialization
> * **Multistream**: Self-describing stream network protocols

Reviewer: ```suggestion
* **Multibase**: Self-describing base encodings
* **Multihash**: Self-describing hashes
* **Multicodec**: Self-describing serialization
* **Multiaddr**: Self-describing network addresses
* **Multistream**: Self-describing stream network protocols
```

just a minor shuffle to help with relevance to IPLD, not particularly important but multiaddr, multistream and multigram are unrelated to IPLD & CIDs so are adjacent concerns.

The different hash functions can be seen in the [multicodec table](https://github.com/multiformats/multicodec)). For example, the SHA3-384 hash function has the code `0x15`, while a BLAKE2b hash function with 384-bit output is `0xb230`. After the hash function code prefix, a Multihash also prefixes the length of the hash digest to follow. So a Multihash can be said to be: `<hash-function-code><digest-length><digest-bytes>`.

**Multibase** is a protocol that allows for string or text data to self-describes which base encoding is used. Multibase uses a prefix _character_ which describes the base used to encode the bytes that follow. For example:

* `b` - base32
* `z` - base58
* `f` - base16

The full list of bases and their mappings can be found in the Multibase registry at https://github.com/multiformats/multibase

Base32 is the most commonly used base encoding for version 1 CIDs, hence the `b` at the beginning. This character says that all the following characters can be passed through a base32 decoder to retrieve the raw bytes.

> Fun fact: while not the most compact string representation, base32 is the preferred base encoding for CIDv1 because it only uses lower-case ASCII characters so is safe to use in DNS entries (URLs).

Reviewer: ```suggestion
Fun fact: while not the most compact string representation, base32 is the preferred base encoding for CIDv1 because it only uses lower-case ASCII characters so is safe to use in DNS entries (and therefore URLs).
```

### Interpreting a CID

> A CID is the concatenation of the multiformats `<multibase>(<cid-version><multicodec><multihash>)`. When represented as a string, the Multibase character is prefixed, but when represented as bytes we don't need Multibase so it's simply `<cid-version><multicodec><multihash>`.

Reviewer: ```suggestion
A CID is the concatenation of multiformat values `<multibase>(<cid-version><multicodec><multihash>)`. When represented as a string, the Multibase character is prefixed, but when represented as bytes we don't need Multibase so it's simply `<cid-version><multicodec><multihash>`.
```
